RFC-0001-torch-function-for-methods.md

```diff
@@ -0,0 +1,177 @@
+
+```python
+class SubTensor(torch.Tensor):
+    def __torch_tensor__(self, func, types, args, kwargs):
+        # Implementation here
+```
+
+The reason for adding `types` to the signature is necessitated by the need for `super()`. If we set a requirement for `super().__array_function__` to work properly, we would need to provide an easy way for users to signal to `__array_function__` that they are calling to the next-specific implementation. The way we propose to handle this is the same as it is handled in NumPy, albeit not in the context of overriding methods, but rather, in the context of subclasses of `numpy.ndarray` or other classes that implement `__array_function__`.
+
+The process followed during a function/method call would be equivalent to:
+
+1. The dispatcher is called to extract the `Tensor`-likes.
+2. All `Tensor`-likes are checked for `__torch_function__`. If none exist, the internal implementation is called, and the final result is returned.
+3. A collection of types that implement `__torch_function__` is created, with no guaranteed order other than that subclasses come before superclasses.
+4. For one argument of each type, `__torch_function__` is called. The first such function or method to return something other than `NotImplemented` will be the final result. All exceptions will be propagated upward.
+5. If all `__torch_function__` implementations return `NotImplemented`, a `TypeError` is raised with an appropriate error message.
+
+In practice, for most PyTorch functions, the list of tensor-likes is already available and the dispatcher doesn't need to be called. Additionally, while equivalent to the code above, if the `Tensor`-likes are all `Tensor`s or classes for which `__torch_function__ is torch.Tensor.__torch_function__`, the internal implementation is called immediately. This is done as a performance optimisation to avoid overhead for concrete `Tensor` objects.
+
+To access super, one would do the following:
+```python
+class SubTensor(torch.Tensor):
+    def __torch_tensor__(self, func, types, args, kwargs):
+        # Pre-processing here
+        val = super().__torch_function__(
+            func,
+            tuple(
+                t for t in types if not issubclass(t, SubTensor)
+            ),
+            args,
+            kwargs
+        )
+        # Post processing here
+```
+
+This way `__torch_function__` knows the list of types to dispatch to, and it will _not_ dispatch to `SubTensor` anymore in this example.
+
+We will also recommend that all `Tensor` subclasses make their own methods go through `__torch_function__` via a decorator `@torch_function_dispatch`. This decorator was added and then removed for performance reasons, however it will be added back to allow external libraries to interface with the protocol. It will take a single argument: a dispatcher, i.e. a callable that returns an iterable of all the "duck-Tensors", or possible candidates for classes that may implement `__torch_function__`.
```

Comment:
That `torch_function_dispatch` addition should be added as a main point in your proposal (point 5 in Motivation and Scope section). Especially because you not only want to add it back, but to make it public (NumPy hasn't even done that yet....).